Added vec_clear and vec_find. Implemented lazy init

juliuskoskela opened the pull request:
- Implemented lazy initialization for the vector. Now if user passes init_len == 0 to `vec_new`, memory won't be allocated at this point, but when the vector is first used (push, insert, etc).
- Implemented `vec_clear` that just changes the length of the vector to zero reserving all allocated space.
- Implemented `vec_find` a simple search taking `f` comaprison method as arg.

Commit:
Added vec_clear and vec_find. Implemented lazy init

satukoskinen committed:
Change tabs to spaces

satukoskinen commented:
I'll split my notes into sections:

1) Inspecting the changes might be easier if changes would be grouped to commits that would be independent of one another. I'd recommend that you try out making use of the VSCode Source Control to make partial commits from a file (only stage a specific section of a file), if needed.

2) Vec_clear and vec_find

`vec_clear`: do you need to allow for the user to pass in a function to free allocated memory from within the elements that are "cleared"?

3) Lazy initialization: looks good

4) Other functions

`vec_free`: should this incorporate the iteration of elements to free whatever might need to be freed from within them? is it the user's responsibility to iterate the array to deallocate the memory allocated within the elements?

`vec_get`: should not return an index that does not contain anything, I think. here, it would be when index == src->len. 

`vec_from`: `dst->alloc_size` bytes is copied regardless of how many elements there are in `src`. Will garbage get copied? What if src is empty? There should be a check for what the right copy size is.

`vec_copy`: same issue as above, `copy_size` is set to `src->alloc_size`. Should it be `dst->alloc_size`? Also, if dst->memory is null, a new vector is created but the size is set confusingly with `src->len` and `dst->elem_size`. Should the latter be `src->elem_size`?

`vec_resize`: is the call to `vec_copy` redundant?

`vec_push`: could the argument in `vec_resize` simply be dst->len * 2?

`vec_remove`: would you not want to implement same functionality as in `vec_pop`? the user can just throw the pointer away if that's not needed. Also, if you attempt vec_remove on a vector with length of 0 with index 0, `len` will wrap around.

`vec_append` and `vec_prepend`: are `1` and `dst->elem_size` the correct one to use if `dst->memory` does not exist? I don't quite understand the allocation strategy. also, if you do changes for `vec_get`, that has implications here.

Some of these may be misunderstandings related to lazy initialization, though.

Commit:
Corrected vec_get to return NULL from index = src->len

Commit:
Tabs...

Commit:
Corrected copy sizes in vec_copy, removed redundant vec_copy from vec_resize, corrected resizing amount in vec_push

juliuskoskela commented:
1. You are correct.

2. This is a good question. Let's think about it separately.

4.1 I think it's ok to assume the user will "manually" manage possible data deallocation in the vector.

```c
void deallocator(void *elem)
{
    // Element deallocation.
}

int main(void)
{
    t_vec v;
    ...
    vec_iter(&v, deallocator);
    vec_free(&v);
    ....
}
```

I see the main use-case for the vector being with unallocated data anyways. However a `vec_destructor(t_vec *src, (*dealloc)(void *))` could possibly be implemented as well.

4.2 Vec from is meant to be a constructor, so it allocates a new vector from the contents found in the pointer. In this case the alloc_size will always match.

4.3 This is correct I think. We either copy as much as is available in src or if dst->alloc_size can't fit everything we copy as much as we can ie. dst->alloc_size amount. There is the question should we actually make an error condition for `src->elem_size != dst->elem_size`. It's a not so obious if we want to do that. If not the data could be sort of "typecast" by copying from a vec of other type to another. Don't know if this is important though.

4.4 Indeed it is, corrected.

4.5 Also corrected.

4.6 Will think on this in a separate PR

4.7 These are also "optimizations" I will consider separately.

Commit:
Don't ask...

Commit:
tabs..
